python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -235,14 +245,205 @@ private Node update(Node node) {
+ * adding a dataflow step from the dictionary argument to `TKwOverflowNode(call, callable)` and a
+ * step to clear content of that node at any unpacked keys.
+ *
+ * ## Examples:
+ * Assume that we have the callable
+ * ```python
+ * def f(x, y, *t, *d):
+ *   pass
+ * ```
+ * Then the call
+ * ```python
+ * f(0, 1, 2, a=3)
+ * ```
+ * will be modelled as
+ * ```python
+ * f(0, 1, [*t], [**d])
+ * ```
+ * where `[` and `]` denotes synthesisezed nodes, so `[*t]` is the synthesized tuple argument
+ * `TPosOverflowNode` and `[**d]` is the synthesized dictionary argument `TKwOverflowNode`.
+ * There will be a store step from `2` to `[*t]` at pos `0` and one from `3` to `[**d]` at key
+ * `a`.
+ *
+ * For the call
+ * ```python
+ * f(0, **{"y": 1, "a": 3})
+ * ```
+ * no tuple arguemnt is synthesized. It is modelled as
+ * ```python
+ * f(0, [y=1], [**d])
+ * ```
+ * where `[y=1]` is the synthesised unpacked argument `TKwUnpacked` (with `name` = `y`). There is
+ * a read step from `**{"y": 1, "a": 3}` to `[y=1]` at key `y` to get the value passed to the parameter
+ * `y`. There is a dataflow step from `**{"y": 1, "a": 3}` to `[**d]` to transfer the content and
+ * a clearing of content at key `y` for node `[**d]`, since that value has been unpacked.
+ */
+module ArgumentPassing {
+  /**
+   * Holds if `call` is a call to `callable`.
+   * Used to limit the size of predicates.
+   */
+  predicate connects(CallNode call, CallableValue callable) {
+    exists(DataFlowCall c |
+      call = c.getNode() and
+      callable = c.getCallable().getCallableValue()
+    )
+  }
+
+  /**
+   * Gets the `n`th parameter of `callable`.
+   * If the callable has a starred parameter, say `*tuple`, that is matched with `n=-1`.
+   * If the callable has a doubly starred parameter, say `**dict`, that is matched with `n=-2`.
+   * Note that, unlike other languages, we do _not_ use -1 for the position of `self` in Python,
+   * as it is an explicit parameter at position 0.
+   */
+  NameNode getParameter(CallableValue callable, int n) {
+    // positional parameter
+    result = callable.getParameter(n)
+    or
+    // starred parameter, `*tuple`
+    exists(Function f |
+      f = callable.getScope() and
+      n = -1 and
+      result = f.getVararg().getAFlowNode()
+    )
+    or
+    // doubly starred parameter, `**dict`
+    exists(Function f |
+      f = callable.getScope() and
+      n = -2 and
+      result = f.getKwarg().getAFlowNode()
+    )
+  }
+
+  /**
+   * Gets the argument to `call` that is passed to the `n`th parameter of `callable`.
+   * If it is a positional argument, it must appear at position `argNr`.
+   * `argNr` will differ from `n` for method- or class calls, where the first parameter
+   * is `self` and the first positional argument is passed to the second positional parameter.
+   */
+  Node getArg(CallNode call, int argNr, CallableValue callable, int n) {
```

Comment:
I'm not a big fan of the naming of `argNr`, and it took me some time to figure out what it meant. Here is a solution that is a bit ugly, but it's also very clear what is going on.

```suggestion
  /**
   * Gets the argument to `call` that is passed to the `paramN`th parameter of `callable`.
   * If it is a positional argument, it is the `argN`'th argument of `call`.
   * `argN` will differ from `paramN` for method- or class calls, where the first parameter
   * is `self` and the first positional argument is passed to the second positional parameter.
   */
  Node getArg(CallNode call, int argNr, CallableValue callable, int paramN) {
```

I'm also not entirely sure what a "class call" is. Is it the call to `__init__`? or is it a call to a classmethod?

Reply:
I had `argNr` be an offset at one point, but that did not seem clear. Then I thought it had to be `argNr` and `paramNr` to make the distinction clear in the readers head from the get-go. Then, after having switched to `argNr`, I felt that `n` was fine given the comment about the `n`th parameter. Clearly that was not true :-)

Reply:
A class call is `C()` for some class `C`, which we consider a call to `C.__init__`.

Reply:
I have tried the version with `argNr` and `paramNr`, please let me know if that is clearer...

Reply:
Nice, think it's an improvement for sure :muscle: 

Two things though:

### 1) classmethod

What about classmethod calls? They also take a "special" first argument.

```py
class Foo:
    @classmethod
    def func(cls):
        print(cls)
Foo.func()
```

### 2) naming for `Nr`

secondly, I'm not a big fan of the `Nr` naming scheme. That leads me to think _number_, and then I start getting unsure whether using `paramNr = 0` or `paramNr = 1` will give me the first parameter of the function. Can we stick to using `N` (without the `r`) and instead of talking about positions, talk about indexes (possibly explaining they are zero-based, although I don't care too strongly about that).

If you're not feeling like doing this, I would be very happy to be enlightened as to why `paramNr` is clearly better :wink:

Reply:
We try the one with `argN` and `paramN`, it actually does not look too bad to me now..
Class methods have been somewhat overlooked 😮, they are treated as methods (flow is found on lines 80 and 81 in `data model.py` but not on line 82), so the first argument will be wrong. I added a TODO.